File: src/contests/cg/puzzle/lua/euclids-algorithm.lua
-- 🎮 CodinGame Puzzle - euclids-algorithm
-- https://www.codingame.com/training/easy/euclids-algorithm

local sa, sb = io.read():match("(%S+)%s+(%S+)")
local a, b = tonumber(sa), tonumber(sb)
local x, y = a, b
while y ~= 0 do
  local q = x // y
  local r = x % y
  print(string.format("%d=%d*%d+%d", x, y, q, r))
  x, y = y, r
end
print(string.format("GCD(%d,%d)=%d", a, b, x))

File: src/contests/cg/puzzle/haskell/euclids-algorithm.hs
-- 🎮 CodinGame Puzzle - euclids-algorithm
-- https://www.codingame.com/training/easy/euclids-algorithm

main :: IO ()
main = do
  line <- getLine
  let [a, b] = map read (words line) :: [Integer]
      go x y
        | y == 0 = putStrLn ("GCD(" ++ show a ++ "," ++ show b ++ ")=" ++ show x)
        | otherwise = do
            let q = x `div` y
                r = x `mod` y
            putStrLn (show x ++ "=" ++ show y ++ "*" ++ show q ++ "+" ++ show r)
            go y r
  go a b
